### PR TITLE
Partial assembly based residual distribution method

### PR DESCRIPTION
**Pull request by artv3**
This PR does a refactoring to move towards the MF Residual Distribution method. 

At the moment we are now able to use the new MFEM 4+ routines to compute geometric factors/face normals etc... and evaluate the action of the face terms.  

Goals of this PR are: 
- [x] Use new MFEM PA/GPU friendly routines to compute the low order solution 
- [x] Create MF PA based 2D methods for Residual Distribution
- [x] Create MF PA based 3D methods for Residual Distribution
- [x] Clean up pass....

### CODE REVIEW

**Comment by artv3** [comment]
@vladotomov  I think this is ready for prime time! The kernel's aren't quite optimized yet but we are now have the option of a fully matrix free remap framework.

File: remhos.cpp [hunk 1]
```diff
@@ -600,9 +617,26 @@ int main(int argc, char *argv[])
    }
    else if (lo_type == LOSolverType::ResDist)
    {
-      const bool subcell_scheme = false;
-      lo_solver = new ResidualDistribution(pfes, k, asmbl, lumpedM,
-                                           subcell_scheme, time_dep);
+     const bool subcell_scheme = false;
+     if(pa)
```
Comment (tzanio) [inline remhos.cpp:621]: Styling comment -- here and below, please leave a space between `if` and `(`:

```suggestion
     if (pa)
```

File: remhos.cpp [hunk 2]
```diff
@@ -633,11 +667,11 @@ int main(int argc, char *argv[])
       s.ProjectCoefficient(sc);
 
       us.MakeRef(&pfes, S, offset[1]);
-      us.HostWrite();
-      u.HostRead();
-      s.HostRead();
+      double *h_us = us.HostWrite();
+      const double *h_u  = u.HostRead();
+      const double *h_s  = s.HostRead();
```
Comment (tzanio) [inline remhos.cpp:672]:
```suggestion
      const double *h_u = u.HostRead();
      const double *h_s = s.HostRead();
```

File: remhos_lo.cpp [hunk 3]
```diff
@@ -238,4 +238,712 @@ void ResidualDistribution::CalcLOSolution(const Vector &u, Vector &du) const
...
+   const DofToQuad *maps = &el_trace->GetDofToQuad(*ir, DofToQuad::TENSOR);
+
+   const int dim = pmesh->Dimension();
+   quad1D = maps->nqpt;
+   dofs1D = maps->ndof;
+   if (dim == 2) { face_dofs = quad1D; }
+   if (dim == 3) { face_dofs = quad1D*quad1D; }
+}
+
+//Taken from DGTraceIntegrator::SetupPA L:145
```
Comment (tzanio) [inline remhos_lo.cpp:265]:
```suggestion
// Taken from DGTraceIntegrator::SetupPA L:145
```

**Comment by tzanio** [comment]
Mention in `CHANGELOG`?

**Review by vladotomov — APPROVED** [review]

**Pull request merged**